### PR TITLE
mgr/cephadm: auto-enable mirroring module when deploying service

### PR DESCRIPTION
**Pull request by phlogistonjohn**
Automatically enable the mgr's mirroring module when creating
cephfs-mirror services. This will trigger a mgr respawn.

Fixes: https://tracker.ceph.com/issues/50593

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by phlogistonjohn** [comment]
@adk3798, @sebastian-philipp  I looked around for existing tests and even docs to update if needed but I couldn't turn anything up. But I admit it's possible I overlooked something. Please point me in the right direction if I did overlook something.

If not, I'd be happy to follow up on this with new tracker tickets to add tests & etc but I'll probably need a little hand-holding with that process too.

**Review by mgfritch — APPROVED** [review]

File: src/pybind/mgr/cephadm/services/cephadmservice.py [hunk 1]
```diff
@@ -986,6 +986,18 @@ def prepare_create(self, daemon_spec: CephadmDaemonDeploySpec) -> CephadmDaemonD
...
     TYPE = 'cephfs-mirror'
 
+    def config(self, spec: ServiceSpec) -> None:
+        # make sure mirroring module is enabled
+        mgr_map = self.mgr.get('mgr_map')
+        mod_name = 'mirroring'
+        if mod_name not in mgr_map.get('services', {}):
+            self.mgr.check_mon_command({
+                'prefix': 'mgr module enable',
+                'module':  mod_name
```
Comment (mgfritch) [inline src/pybind/mgr/cephadm/services/cephadmservice.py:996]:
```suggestion
                'module': mod_name
```

cephadm/services/cephadmservice.py:996:26: E241 multiple spaces after ':'
1     E241 multiple spaces after ':'

**Comment by mgfritch** [comment]
> @adk3798, @sebastian-philipp I looked around for existing tests and even docs to update if needed but I couldn't turn anything up. But I admit it's possible I overlooked something. Please point me in the right direction if I did overlook something.
> 

I don't believe we have existing tests or doc for this. Adding a mirroring test would be nice, but probably not worthwhile for this PR.

**Comment by sebastian-philipp** [comment]
> > @adk3798, @sebastian-philipp I looked around for existing tests and even docs to update if needed but I couldn't turn anything up. But I admit it's possible I overlooked something. Please point me in the right direction if I did overlook something.
> 
> I don't believe we have existing tests or doc for this. Adding a mirroring test would be nice, but probably not worthwhile for this PR.

I mean why not? The test is super simple:

```diff
diff --git a/src/pybind/mgr/cephadm/tests/test_services.py b/src/pybind/mgr/cephadm/tests/test_services.py
index d023369a2ee..990472709bd 100644
--- a/src/pybind/mgr/cephadm/tests/test_services.py
+++ b/src/pybind/mgr/cephadm/tests/test_services.py
@@ -777,3 +777,15 @@ class TestIngressService:
                 }
 
                 assert haproxy_generated_conf[0] == haproxy_expected_conf
+
+
+class TestCephFsMirror:
+    @patch("cephadm.serve.CephadmServe._run_cephadm")
+    def test_config(self, _run_cephadm, cephadm_module: CephadmOrchestrator):
+        _run_cephadm.side_effect = async_side_effect(('{}', '', 0))
+        with with_host(cephadm_module, 'test'):
+            with with_service(cephadm_module, ServiceSpec('cephfs-mirror')):
+                cephadm_module.assert_issued_mon_command({
+                    'prefix': 'mgr module enable',
+                    'module': 'mirroring'
+                })
```

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2022-01-17_12:42:04-orch:cephadm-wip-swagner-testing-2022-01-17-1014-distro-default-smithi/

* ingress is broken
* hardware issue on smithi187
* crun + cgroup issue

**Comment by phlogistonjohn** [comment]
How would you prefer I move forward on this?  I have no strong feelings about things. I can follow up with a separate PR but I want to make sure we're all on the same page. :-)

**Comment by sebastian-philipp** [comment]
> but probably not worthwhile for this PR.

@mgfritch ok after doing a bunch of backports, one thing is obvious: we're getting an increasing amount of non-trivial conflicts. Having a pytest for each PR *really* speeds up backporting now.

**Comment by mgfritch** [comment]
> > but probably not worthwhile for this PR.
> 
> @mgfritch ok after doing a bunch of backports, one thing is obvious: we're getting an increasing amount of non-trivial conflicts. Having a pytest for each PR **really** speeds up backporting now.

agree, we might as well add it to this PR since you've basically written the test already :)

**Pull request merged**